modules/drivers/audio/respeaker.cc

```diff
@@ -0,0 +1,156 @@
+/******************************************************************************
+ * Copyright 2017 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/drivers/audio/respeaker.h"
+
+namespace apollo {
+namespace drivers {
+namespace audio {
+
+// Helper functions
+void report_error(PaError err, const std::string &func_name) {
+  AERROR << "an error occured while calling " << func_name;
+  AERROR << "error number: " << err;
+  AERROR << "error message: " << Pa_GetErrorText(err);
+  throw std::runtime_error("");
+}
+
+// Stream
+Stream::~Stream() { Pa_CloseStream(pastream_ptr_); }
+
+void Stream::init_stream(int rate, int channels, int chunk,
+                         int input_device_index, PaSampleFormat format) {
+  PaError err;
+
+  // Init parameters of input device
+  inputParameters_ptr_ = reinterpret_cast<PaStreamParameters *>(
+      malloc(sizeof(PaStreamParameters)));
+  inputParameters_ptr_->device = input_device_index;
+  inputParameters_ptr_->channelCount = channels;
+  inputParameters_ptr_->sampleFormat = format;
+  inputParameters_ptr_->suggestedLatency =
+      Pa_GetDeviceInfo(inputParameters_ptr_->device)->defaultLowInputLatency;
+  inputParameters_ptr_->hostApiSpecificStreamInfo = nullptr;
+
+  err = Pa_OpenStream(
+      &pastream_ptr_, inputParameters_ptr_, nullptr, rate, chunk,
+      paClipOff,  // we only use input so don't bother clipping them *
+      nullptr, nullptr);
+  if (err != paNoError) report_error(err, "Pa_OpenStream");
+  err = Pa_StartStream(pastream_ptr_);
+  if (err != paNoError) report_error(err, "Pa_StartStream");
+}
+
+void Stream::read_stream(int n_frames, char *buffer) {
+  int err =
+      Pa_ReadStream(pastream_ptr_, reinterpret_cast<void *>(buffer), n_frames);
+  if (err != paNoError) {
+    report_error(err, "Pa_ReadStream");
+  }
+}
+
+int Stream::get_chunk_size(int n_frames) {
+  return (n_frames) * (inputParameters_ptr_->channelCount) *
+         (Pa_GetSampleSize(inputParameters_ptr_->sampleFormat));
+}
+
+// Respeaker
+Respeaker::~Respeaker() { Pa_Terminate(); }
+void Respeaker::init(std::shared_ptr<SpeakerConfig> speaker_config) {
```

Comment:
Better use `const std::shared_ptr<SpeakerConfig>&`  since `speaker_config` was read only in the `init` function.

Reply:
You are right but I think it should be `const std::shared_ptr<const SpeakerConfig>&` which also prevents modification to `SpeakerConfig` itself, while `const std::shared_ptr<SpeakerConfig>&` just prevents the re-assignment of that shared_ptr variable?